### PR TITLE
[XPU][DeepSeekV4]Add Deepseek-v4  fused_indexer_q_rope_quant fp8 and mxfp4 kernel SYCL support

### PR DESCRIPTION
**Pull request by jjmiao1**
# Add DeepSeek-V4 Fused Indexer Q + RoPE SYCL Kernels

## Purpose

Add SYCL implementations for the DeepSeek-V4 **fused indexer Q + RoPE** kernels used in the Q-projection stage of the MLA sparse attention pipeline. This PR includes two kernels:

1. **`deepseek_fused_indexer_q_rope_fp8`** — Fused indexer Q + RoPE + FP8 E4M3 quantization
2. **`deepseek_fused_indexer_q_rope_mxfp4`** — Fused indexer Q + RoPE + MXFP4 quantization

Both kernels fuse the following operations into a single pass:
- **Index select:** Extract the sparse Q heads from the full Q projection
- **RoPE rotation:** GPT-J style RoPE on the rope dimensions (dim 64–127)
- **Quantization:** Per-head FP8 or per-block MXFP4 quantization with power-of-2 scales
- **Output layout:** NoPE dimensions straight-quantized, RoPE dimensions interleaved `[re, ro]` pairs

---

### Part 1: `deepseek_fused_indexer_q_rope_fp8`

**Commit:** `b3446a4` — Add Deepseek-v4 fused_indexer_q_rope_fp8 sycl kernel

#### Files Changed

| File | Change |
|------|--------|
| `csrc/xpu/sycl/deepseek_fused_indexer_q_rope_fp8.cpp` | New SYCL kernel (218 lines) |
| `csrc/xpu/ops.h` | Function declaration |
| `csrc/xpu/torch_bindings.cpp` | Op registration (`def` + `impl`) |
| `CMakeLists.txt` | Add source to build |
| `tests/test_deepseek_fused_indexer_q_rope_fp8.py` | Pytest with parametrized correctness + semantic tests |

#### Design

- Grid: one subgroup (SG_SIZE=16) per `(token, head_group)`, HEADS_COARSEN=2
- Per-head: load 128 bf16 → apply RoPE on dim[64:127] → absmax reduce → pow2 scale → FP8 E4M3 quantize
- Branchless FP8 conversion (bitmask select, no divergent branches)
- Registered as `torch.ops._xpu_C.deepseek_fused_indexer_q_rope_fp8`

#### Unit Test — All PASSED

```bash
pytest tests/test_deepseek_fused_indexer_q_rope_fp8.py -v
```

- `test_correctness`: num_tokens ∈ {1, 64, 128} × seed ∈ {0, 42}
- `test_weight_fold_semantics`: validates indexer weight folding correctness

#### Kernel Benchmark: SYCL vs Triton

Platform: Intel Arc Pro B60 (HBM peak bandwidth: 456 GB/s), Warmup 50 iters, Timed 200 iters

| num_tokens | triton(us) | sycl(us) | triton BW(GB/s) | sycl BW(GB/s) | MBU | speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 53.20 | 41.86 | 0.47 | 0.60 | 0.1% | **1.27x** |
| 4 | 52.96 | 40.93 | 1.90 | 2.47 | 0.5% | **1.29x** |
| 16 | 52.68 | 40.72 | 7.66 | 9.91 | 2.2% | **1.29x** |
| 64 | 53.27 | 40.77 | 30.30 | 39.59 | 8.7% | **1.31x** |
| 128 | 52.80 | 40.87 | 61.12 | 78.99 | 17.3% | **1.29x** |
| 256 | 54.67 | 40.88 | 118.11 | 157.96 | 34.6% | **1.34x** |
| 512 | 75.72 | 52.09 | 170.52 | 247.89 | 54.4% | **1.45x** |
| 1024 | 123.10 | 85.69 | 209.78 | 301.37 | 66.1% | **1.44x** |
| 2048 | 218.33 | 165.11 | 236.61 | 312.87 | 68.6% | **1.32x** |
| 4096 | 434.03 | 309.31 | 238.07 | 333.97 | 73.2% | **1.40x** |
| 8192 | 839.01 | 589.06 | 246.28 | 350.79 | 76.9% | **1.42x** |
| 16384 | 3138.89 | 1146.54 | 131.66 | 360.45 | 79.0% | **2.74x** |

**SYCL achieves 1.27x–2.74x speedup** over Triton. Large token regime: ~73–79% MBU.

#### Standalone SYCL Kernel (device time, event profiling)

Device: Intel Arc Pro B60 (HBM peak bandwidth: 456 GB/s)

| num_tokens | device(us) | BW(GB/s) | MBU |
|---:|---:|---:|---:|
| 1 | 1.77 | 14.27 | 3.1% |
| 4 | 2.29 | 44.10 | 9.7% |
| 16 | 2.29 | 176.39 | 38.7% |
| 64 | 4.47 | 360.99 | 79.2% |
| 128 | 8.22 | 392.97 | 86.2% |
| 256 | 14.87 | 434.19 | 95.2% |
| 512 | 29.64 | 435.72 | 95.6% |
| 1024 | 65.21 | 396.11 | 86.9% |
| 2048 | 145.39 | 355.31 | 77.9% |
| 4096 | 284.65 | 362.97 | 79.6% |
| 8192 | 568.46 | 363.50 | 79.7% |
| 16384 | 1130.17 | 365.67 | 80.2% |

Peak device MBU: **95.6%** at 512 tokens. Large token regime stabilizes at ~80% MBU.

#### End-to-End Model Benchmark (FP8 path)

**Note on E2E setup:** Due to current hardware resource constraints, we use a **10-layer trimmed checkpoint** of DeepSeek-V4-Flash-Base for E2E validation. This is sufficient to verify correct integration and no performance regression, as the kernel executes identically in every decoder layer.

**Integration:** In `vllm/models/deepseek_v4/common/ops/fused_indexer_q.py`, the SYCL path calls `torch.ops._xpu_C.deepseek_fused_indexer_q_rope_fp8` to replace the original Triton `fused_indexer_q_rope_quant` implementation.

Model: DeepSeek-V4-Flash-Base (**10 layers**, trimmed), TP=4, input_len=2048, output_len=1024, max_concurrency=1, num_prompts=5, num_warmups=2

| Metric | Triton (baseline) | SYCL | Delta |
|--------|---:|---:|---:|
| Output token throughput (tok/s) | 8.44 | 8.46 | +0.2% |
| Total token throughput (tok/s) | 25.32 | 25.39 | +0.3% |
| Mean TTFT (ms) | 2241.94 | 2244.87 | +0.1% |
| Median TTFT (ms) | 2244.76 | 2244.29 | 0.0% |
| P99 TTFT (ms) | 2250.69 | 2254.80 | +0.2% |
| Mean TPOT (ms) | 116.42 | 116.08 | **-0.3%** |
| Median TPOT (ms) | 116.46 | 115.98 | **-0.4%** |
| P99 TPOT (ms) | 116.69 | 116.86 | +0.1% |

E2E results confirm **no performance regression**. The fused_indexer_q kernel is a lightweight op in the overall decode pipeline (device time ~2us at decode batch size), so its E2E impact is negligible. TPOT and TTFT are within noise range, indicating correct integration with no overhead.

---

### Part 2: `deepseek_fused_indexer_q_rope_mxfp4`

**Commit:** `4da7c4a` — Add Deepseek-v4 fused_indexer_q_rope_mxfp4 sycl kernel

#### Files Changed

| File | Change |
|------|--------|
| `csrc/xpu/sycl/deepseek_fused_indexer_q_rope_mxfp4.cpp` | New SYCL kernel (242 lines) |
| `csrc/xpu/ops.h` | Function declaration |
| `csrc/xpu/torch_bindings.cpp` | Op registration (`def` + `impl`) |
| `CMakeLists.txt` | Add source to build |
| `tests/test_deepseek_fused_indexer_q_rope_mxfp4.py` | Pytest with correctness + edge case tests |

#### Design

- Grid: one subgroup (SG_SIZE=16) per `(token, head_group)`, HEADS_COARSEN=4
- Per-head: 128 bf16 split into 4 MXFP4 blocks (32 elements each)
- Each block: absmax reduce → pow2 scale → nibble-pack to 4-bit output
- Uses `sycl::reduce_over_group` for efficient sub-group reduction
- Registered as `torch.ops._xpu_C.deepseek_fused_indexer_q_rope_mxfp4`

#### Unit Test — All PASSED

```bash
pytest tests/test_deepseek_fused_indexer_q_rope_mxfp4.py -v
```

- `test_correctness`: num_tokens ∈ {1, 4, 16} × seed ∈ {42}
- `test_edge_cases`: distribution ∈ {uniform_50, tiny, zeros, mixed}

#### Kernel Benchmark: SYCL vs Triton

Platform: Intel Arc Pro B60 (HBM peak bandwidth: 456 GB/s), Warmup 50 iters, Timed 200 iters

> Note: Triton MXFP4 implementation relies on CUDA PTX inline asm; comparison here uses Triton's XPU fallback path.

| num_tokens | triton(us) | sycl(us) | triton BW(GB/s) | sycl BW(GB/s) | MBU | speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 83.70 | 69.85 | 0.26 | 0.31 | 0.1% | **1.20x** |
| 4 | 83.80 | 68.81 | 1.02 | 1.24 | 0.3% | **1.22x** |
| 16 | 83.33 | 68.38 | 4.10 | 5.00 | 1.1% | **1.22x** |
| 64 | 83.75 | 68.62 | 16.34 | 19.95 | 4.4% | **1.22x** |
| 128 | 83.62 | 68.53 | 32.73 | 39.94 | 8.8% | **1.22x** |
| 256 | 83.54 | 68.64 | 65.50 | 79.75 | 17.5% | **1.22x** |
| 512 | 119.77 | 68.89 | 91.37 | 158.92 | 34.9% | **1.74x** |
| 1024 | 193.80 | 91.21 | 112.95 | 240.07 | 52.6% | **2.12x** |
| 2048 | 345.51 | 161.35 | 126.73 | 271.43 | 59.5% | **2.14x** |
| 4096 | 644.72 | 287.89 | 135.82 | 304.25 | 66.7% | **2.24x** |
| 8192 | 1251.79 | 531.45 | 139.90 | 329.62 | 72.3% | **2.35x** |
| 16384 | 2773.88 | 1027.27 | 126.28 | 341.05 | 74.8% | **2.70x** |

**SYCL achieves 1.20x–2.70x speedup** over Triton. Large token regime: ~67–75% MBU.

#### Standalone SYCL Kernel (device time, event profiling)

Device: Intel Arc Pro B60 (HBM peak bandwidth: 456 GB/s)

| num_tokens | device(us) | BW(GB/s) | MBU |
|---:|---:|---:|---:|
| 1 | 3.00 | 7.1 | 1.6% |
| 4 | 3.09 | 27.6 | 6.1% |
| 16 | 3.22 | 106.4 | 23.3% |
| 64 | 4.72 | 289.8 | 63.6% |
| 128 | 8.78 | 311.8 | 68.4% |
| 256 | 16.23 | 337.2 | 73.9% |
| 512 | 29.36 | 372.9 | 81.8% |
| 1024 | 58.17 | 376.5 | 82.6% |
| 2048 | 130.19 | 336.4 | 73.8% |
| 4096 | 250.78 | 349.3 | 76.6% |
| 8192 | 495.22 | 353.7 | 77.6% |
| 16384 | 986.69 | 355.1 | 77.9% |

Peak device MBU: **82.6%** at 1024 tokens. Large token regime stabilizes at ~77–78% MBU.

#### End-to-End Model Benchmark

> The current model pipeline uses FP8 fused_indexer_q only. MXFP4 path is not ready yet so there is no E2E model benchmark results.

### CODE REVIEW

**Comment by jjmiao1** [comment]
@jikunshang please help review, thanks

File: csrc/xpu/sycl/deepseek_fused_indexer_q_rope_fp8.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+#include <sycl/sycl.hpp>
+#include "utils.h"
+#include <cstdint>
+#include <torch/all.h>
+
+namespace vllm {
+
+using bf16_t = sycl::ext::oneapi::bfloat16;
+
+static constexpr int HEAD_DIM = 128;
```
Comment (mayuyuace) [inline csrc/xpu/sycl/deepseek_fused_indexer_q_rope_fp8.cpp]: So the kernel is only for deepseek-V4.
Is there possible that other models use this kernel, for example subsequent deepseek models?
If that, it is better to use template to dispatch these params instead of hard code.

**Review by wenbinc-Bin — APPROVED** [review]

File: csrc/xpu/sycl/deepseek_fused_indexer_q_rope_mxfp4.cpp [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    uint32_t data[HEADS_COARSEN][4];
+    #pragma unroll
+    for (int hc = 0; hc < HEADS_COARSEN; ++hc) {
+      const uint32_t* q32 = reinterpret_cast<const uint32_t*>(
+          q_ + tok * qs0 + (bh + hc) * HEAD_DIM);
+      data[hc][0] = q32[lane];
+      data[hc][1] = q32[HALF_BLOCK + lane];
+      data[hc][2] = q32[NOPE_DIM / 2 + lane];
+      data[hc][3] = q32[NOPE_DIM / 2 + HALF_BLOCK + lane];
+    }
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply csrc/xpu/sycl/deepseek_fused_indexer_q_rope_mxfp4.cpp:119]: Already checked before kernel launch `TORCH_CHECK(num_heads % 4 == 0, "num_heads must be divisible by 4");`

**Review by Wei-Lin-Intel — APPROVED** [review]
LGTM

File: csrc/xpu/ops.h [hunk 3]
```diff
@@ -154,4 +154,25 @@ torch::Tensor fp8_paged_mqa_logits(
...
+void deepseek_fused_indexer_q_rope_mxfp4(
+    const torch::Tensor& q,
+    const torch::Tensor& positions,
+    const torch::Tensor& cos_sin_cache,
+    const torch::Tensor& index_weights,
+    double softmax_scale,
+    double head_scale,
+    torch::Tensor& packed_out,
+    torch::Tensor& scales_out,
+    torch::Tensor& weights_out);
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply csrc/xpu/ops.h]: These are guarded at the CMakeList with `XPU_SPECIFIC_KERNELS_ENABLED` which also guards ` torch_bindings.cpp`, when the option is OFF, neither the bindings nor the kernels will be compiled and no link failure is possible.

File: csrc/xpu/torch_bindings.cpp [hunk 4]
```diff
@@ -135,6 +135,21 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, xpu_ops) {
...
+  xpu_ops.impl("deepseek_fused_indexer_q_rope_fp8", torch::kXPU,
+               &deepseek_fused_indexer_q_rope_fp8);
+
+  xpu_ops.def(
+      "deepseek_fused_indexer_q_rope_mxfp4(Tensor q, Tensor positions, "
+      "Tensor cos_sin_cache, Tensor index_weights, float softmax_scale, "
+      "float head_scale, Tensor! packed_out, Tensor! scales_out, "
+      "Tensor! weights_out) -> ()");
+  xpu_ops.impl("deepseek_fused_indexer_q_rope_mxfp4", torch::kXPU,
+               &deepseek_fused_indexer_q_rope_mxfp4);
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply csrc/xpu/torch_bindings.cpp]: These are guarded at the CMakeList with `XPU_SPECIFIC_KERNELS_ENABLED` which also guards ` torch_bindings.cpp`, when the option is OFF, neither the bindings nor the kernels will be compiled and no link failure is possible.

File: tests/test_deepseek_fused_indexer_q_rope_fp8.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    pytest tests/test_deepseek_fused_indexer_q_rope_fp8.py -v
+"""
+import math
+
+import pytest
+import torch
+
+import vllm_xpu_kernels._xpu_C  # noqa: F401
+
+DEVICE = torch.device("xpu")
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply tests/test_deepseek_fused_indexer_q_rope_fp8.py:17]: No point add extra check since the tests will only run on XPU device.

File: tests/test_deepseek_fused_indexer_q_rope_mxfp4.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    pytest tests/test_deepseek_fused_indexer_q_rope_mxfp4.py -v
+"""
+import math
+
+import pytest
+import torch
+
+import vllm_xpu_kernels._xpu_C  # noqa: F401
+
+DEVICE = torch.device("xpu")
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply tests/test_deepseek_fused_indexer_q_rope_mxfp4.py:17]: No point add extra check since the tests will only run on XPU device.

File: csrc/xpu/sycl/deepseek_fused_indexer_q_rope_mxfp4.cpp [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+      pd[NOPE_DIM / 2 + lane] = p2;
+      pd[NOPE_DIM / 2 + HALF_BLOCK + lane] = p3;
+
+      if (lane == 0) {
+        uint32_t sp = (uint32_t)ue0 | ((uint32_t)ue1 << 8) |
+                      ((uint32_t)ue2 << 16) | ((uint32_t)ue3 << 24);
+        *reinterpret_cast<uint32_t*>(sd) = sp;
+        wo_[tok * H_ + h] =
+            static_cast<float>(w_[tok * H_ + h]) * combined_scale_;
+      }
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply csrc/xpu/sycl/deepseek_fused_indexer_q_rope_mxfp4.cpp:192]: The offset is always a multiple of `NUM_BLOCKS` (4), so alignment is guaranteed.

**Review by mayuyuace — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]
